src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(stateCopy, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach(function (key) {
+        delete stateCopy[key];
+      });
+    }
+
+    if (action.type === 'clear') {
+      for (const key in stateCopy) {
+        if (stateCopy.hasOwnProperty(key)) {
+          delete stateCopy[key];
+        }
+      }
+    }
```

Comment:
[CODE STYLE]: switch/case should always have default case for error handling. Once you refactor the code to use a `switch` statement, remember to include a `default` case to handle any unexpected `action.type` values.